[16.0] [ADD] stock_release_channel_auto_release

lmignon opened the pull request:
* [ ] includes #506
* [ ] depends on #478 
* [ ] depends on https://github.com/OCA/stock-logistics-warehouse/pull/1532
* [ ] depends on https://github.com/OCA/stock-logistics-warehouse/pull/1560
* [ ] depends on https://github.com/OCA/stock-logistics-warehouse/pull/1681

This addon define an automatic release mode on the release channels. By default
release channels manage the release of the transfers in batch mode. This means
that the transfers are released only when the user manually triggers the release
process by clicking on the release button.

When the automatic release mode is enabled, the transfers are released automatically
when a new transfer is added to the channel or as soon a product becomes available.

As for the batch mode, the automatic release process is only active on open channels.
When is locked, the automatic release process is stopped. Once the channel is unlocked,
the automatic release process is restarted and transfers into the channel are released
if they are ready to be released (IOW if quantities are available for moves not
yet released).

lmignon committed:
[IMP] stock_release_channel: Makes assign_release_channel takes only 1 picking as args

Except from the testsuite, the assign_release_channel method from stock.release.channel model is always called with one picking as args. To allows future improvements add a specific contraint on this actual usage

lmignon committed:
[IMP] stock_release_channel: Allows to filter channel candidates into the assign_release_channel method

In some cases, you could need to prevent a channel to be used as candidate into the assign_release_channel mehtod. This change add a hook method that can be used to filter the channels

lmignon committed:
[IMP] stock_release_channel: Manage release channel lifecycle

The availability of a release channel depends on its state. A release channel
can be in one of the following states:

  - Open: the channel is available and can be used to release transfers. New
    transfer are assigned automatically to this channel.
  - Locked: the channel is available but the release of transfers from the channel
    is no more possible. New transfers are still automatically assigned to this
  - Asleep: the channel is not available and cannot be used to release
    transfers. It is also no more possible to assign transfers to this channel.

New release channels are by default "Open". You can change its state by using
the "Lock" and "Sleep" buttons. When the "Sleep" button is used, in addition to
the state change, not completed transfers assigned to the channel are unassigned
from the channel. When the "Lock" button is used, only the state change is changed.
A locked channel can be unlocked by using the "Unlock" button.
A asleep channel can be waken up by using the "Wake up" button. When the "Wake up"
button is used, in addition to the state change, the system looks for pending
transfers requiring a release and try to assign them to a channel in the
"Open" or "Locked" state.

lmignon committed:
[ADD] stock_release_channel_auto_release: Add automatic release mode

This addons define an automatic release mode on the release channels. By default
release channels manage the release of the transfers in batch mode. This means
that the transfers are released only when the user manually triggers the release
process by clicking on the release button.

When the automatic release mode is enabled, the transfers are released automatically
when a new transfer is added to the channel or as soon a product becomes available.

As for the batch mode, the automatic release process is only active on open channels.
When is locked, the automatic release process is stopped. Once the channel is unlocked,
the automatic release process is restarted and transfers into the channel are released
if they are ready to be released (IOW if quantities are available for moves not
yet released).

lmignon commented:
@jbaudoux @sebalix @simahawk 
Tests are failing on `stock_available_to_promise_release` since tests are ran on a DB with the `sale` addon installed but `sale` is not into the dependencies of `stock_available_to_promise_release`. The dependency on the `sale` addon comes from `stock_release_channel`. That's the only addon that depends on `sale`. I've the feeling that this new dependency could make the tests randomly failing as soon as a product is created into a test and the `sale` addon is not yet loaded into the registry....

TDu commented on stock_release_channel_auto_release/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,13 @@
+This addons define an automatic release mode on the release channels. By default
```
```suggestion
This addon defines an automatic release mode on the release channels. By default
```

jbaudoux commented:
> @jbaudoux @sebalix @simahawk Tests are failing on `stock_available_to_promise_release` since tests are ran on a DB with the `sale` addon installed but `sale` is not into the dependencies of `stock_available_to_promise_release`. The dependency on the `sale` addon comes from `stock_release_channel`. That's the only addon that depends on `sale`. I've the feeling that this new dependency could make the tests randomly failing as soon as a product is created into a test and the `sale` addon is not yet loaded into the registry....

I think the dependency on sales is not necessary anymore with my change done here in the meantime: https://github.com/OCA/wms/commit/e334ddc0bece05fd562714205655c809e24a6caa

lmignon commented:
> > @jbaudoux @sebalix @simahawk Tests are failing on `stock_available_to_promise_release` since tests are ran on a DB with the `sale` addon installed but `sale` is not into the dependencies of `stock_available_to_promise_release`. The dependency on the `sale` addon comes from `stock_release_channel`. That's the only addon that depends on `sale`. I've the feeling that this new dependency could make the tests randomly failing as soon as a product is created into a test and the `sale` addon is not yet loaded into the registry....
> 
> I think the dependency on sales is not necessary anymore with my change done here in the meantime: [e334ddc](https://github.com/OCA/wms/commit/e334ddc0bece05fd562714205655c809e24a6caa)

@jbaudoux Tests are now :green_circle:

sbejaoui commented:
@lmignon , @jbaudoux 

https://github.com/acsone/wms/pull/18

jbaudoux commented on stock_release_channel/views/stock_release_channel_views.xml:
```diff
@@ -116,6 +191,7 @@
                 class="oe_background_grey o_kanban_dashboard o_emphasize_colors o_stock_release_channel"
                 create="0"
             >
+                <field name="release_mode" invisible="1" />
```
I don't think you need the invisible in this section

jbaudoux commented on stock_release_channel/views/stock_release_channel_views.xml:
```diff
@@ -148,8 +224,20 @@
                                     <div class="o_kanban_card_header_title">
                                         <div class="o_primary">
                                             <field name="name" />
+                                            <field name="state" invisible="1" />
```
I think it should be outside the template in the list of fields to pre-load in the kanban (without invisible)

jbaudoux replied:
You added it in the header but forgot to remove it from here

jbaudoux commented on stock_release_channel/views/stock_release_channel_views.xml:
```diff
@@ -166,10 +254,15 @@
                                 <div class="container o_kanban_card_content">
                                     <div class="row o_kanban_card_upper_content">
                                         <div class="col-4 o_kanban_primary_left">
+                                            <field
+                                                name="is_release_allowed"
+                                                invisible="1"
+                                            />
```
I think it should be outside the template in the list of fields to pre-load in the kanban (without invisible)

lmignon replied:
done

lmignon committed:
[ADD] stock_release_channel_batch_mode_commercial_partner

This new addon defines the 'Grouped by Commercial Entiry' mode when releasing in batch. This mode was previously defined into the stock_release_channel addon but since it requires to store an additional info on the stock.picking model we moved this functionality in a dedicated addon. This will avoid to create useless additional column on the stock.picking model if we don't need it. Moreover, since having the commercial partner on the stock.picking could make sense outside of the stock_release_channel context, a new addon 'stock_picking_commercial_partner' has been created into the oca/stock-logistics-warehouse repository to provides this generic functionality

jbaudoux reviewed (approved):
Thanks for the last cleanup regarding the commercial partner

rousseldenis commented:
@jbaudoux Could you also review the linked open PR"s? Thanks

rousseldenis commented:
@TDu Could you update your review ?

sbejaoui committed:
[IMP] - stock_release_channel: search by name

rousseldenis commented:
/ocabot merge patch
